doc/conf.py

```diff
@@ -347,12 +348,40 @@
 traceability_json_export_path = '_build/exported_items.json'
 
 def traceability_callback_per_item(name, collection):
+    """Callback function called when an item-directive is being processed.
+
+    Note: attributes, relationships and content (body) of the item can be modified. Sphinx processes each directive
+    in turn, so attributes and relationships added or modified by other directives may not have been processed yet.
+
+    Args:
+        name (str): Name (id) of the item currently being parsed
+        collection (TraceableCollection): Collection of all items that have been parsed so far
+    """
     if name == 'r001':
         item = collection.get_item(name)
         content_str = item.get_content()
         content_str += '\n\nThis line was added by ``traceability_callback_per_item`` and is parsed as |RST| syntax.'
         item.set_content(content_str)
 
+
+def traceability_inspect_item(name, collection):
+    """Callback function called when an item-directive is being rendered.
+
+    Note: the item should not be modified, only inspected. At this stage of the documentation build, all directives,
```

Comment:
Is there any way to enforce this (like `const` in C)

Reply:
The attributes must not be read-only before reaching the stage of rendering. I don't see a clean way of using that `Readonly`.

As an improvement, I delayed the call of this callback function so that any modifications made by it will no longer have an effect on the rendered item: 8c7c74c30e607424188eb3ad7c5534b4724d65cf. This makes it more obvious that it cannot be used to modify the item. The reason that modifications shouldn't be made is because, at this point, all other directive types, e.g. item-attributes-matrix, will already have been rendered.

Reply:
great improvement. Thanks.